File: examples/javascript/meta-agent.js
/**
 * PRISM Framework — Full Pipeline with Meta-Agent (JavaScript)
 *
 * Complete PRISM implementation:
 * 1. N-Sampling: fire N parallel requests
 * 2. Meta-Agent: meritocratic synthesis (not consensus averaging)
 *
 * npm install @anthropic-ai/sdk
 */

import Anthropic from '@anthropic-ai/sdk';
import { readFileSync } from 'fs';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const client = new Anthropic({ apiKey: process.env.ANTHROPIC_API_KEY });
const __dir = dirname(fileURLToPath(import.meta.url));

// Load meta-agent prompt template
const META_PROMPT_TEMPLATE = readFileSync(
  join(__dir, '../../prompts/templates/master-synthesis.md'),
  'utf-8'
);

/**
 * Full PRISM pipeline
 * @param {string} prompt - The task prompt
 * @param {object} options - Pipeline configuration
 * @returns {Promise<{synthesis: string, responses: object[], divergences: string[]}>}
 */
export class PRISMSampler {
  constructor(options = {}) {
    this.n = options.n ?? 5;
    this.model = options.model ?? 'claude-opus-4-6';
    this.temperature = options.temperature ?? 0.9;
    this.maxTokens = options.maxTokens ?? 2000;
    this.metaModel = options.metaModel ?? this.model;
  }

  /**
   * Run the full PRISM pipeline
   */
  async run(prompt, customMetaPrompt = null) {
    console.log(`[PRISM] Starting N-Sampling (n=${this.n})...`);
    const startTime = Date.now();

    // Step 1: Parallel N-Sampling
    const samplingRequests = Array.from({ length: this.n }, (_, i) =>
      client.messages.create({
        model: this.model,
        max_tokens: this.maxTokens,
        temperature: this.temperature,
        messages: [{ role: 'user', content: prompt }],
      }).then(res => ({
        index: i + 1,
        content: res.content[0].text,
      }))
    );

    const responses = await Promise.all(samplingRequests);
    console.log(`[PRISM] ${this.n} responses received (${Date.now() - startTime}ms)`);

    // Step 2: Meta-Agent Synthesis
    console.log(`[PRISM] Running Meta-Agent synthesis...`);

    const responsesFormatted = responses
      .map(r => `### Response ${r.index}\n\n${r.content}`)
      .join('\n\n---\n\n');

    const metaPromptBase = customMetaPrompt || META_PROMPT_TEMPLATE;
    const metaPrompt = `${metaPromptBase}

---

## The ${this.n} responses to synthesize:

${responsesFormatted}`;

    const metaResponse = await client.messages.create({
      model: this.metaModel,
      max_tokens: this.maxTokens * 2,
      temperature: 0.3, // Low temperature for synthesis — we want precision here
      messages: [{ role: 'user', content: metaPrompt }],
    });

    const synthesis = metaResponse.content[0].text;
    const totalTime = Date.now() - startTime;
    console.log(`[PRISM] Synthesis complete (${totalTime}ms total)`);

    return {
      synthesis,
      responses,
      metadata: {
        n: this.n,
        model: this.model,
        temperature: this.temperature,
        totalTimeMs: totalTime,
      },
    };
  }
}

// Example usage
if (process.argv[1] === new URL(import.meta.url).pathname) {
  const prism = new PRISMSampler({ n: 3 }); // n=3 for demo speed

  const prompt = `
    Explain the most underrated recovery technique in strength training,
    with physiological mechanisms and practical implementation.
    Be specific, avoid common knowledge.
  `;

  const result = await prism.run(prompt);

  console.log('\n' + '='.repeat(60));
  console.log('PRISM SYNTHESIS (Meta-Agent Output)');
  console.log('='.repeat(60));
  console.log(result.synthesis);
  console.log('\n' + '='.repeat(60));
  console.log(`Total time: ${result.metadata.totalTimeMs}ms | N: ${result.metadata.n}`);
}

File: examples/python/meta_agent.py
"""
PRISM Framework — Full Pipeline with Meta-Agent (Python)

Complete PRISM implementation:
1. N-Sampling: N parallel async requests
2. Meta-Agent: meritocratic synthesis

pip install anthropic
"""

import asyncio
import os
import time
from pathlib import Path
import anthropic


META_PROMPT_PATH = Path(__file__).parent.parent.parent / "prompts/templates/master-synthesis.md"


class PRISMPipeline:
    def __init__(
        self,
        n: int = 5,
        model: str = "claude-opus-4-6",
        temperature: float = 0.9,
        max_tokens: int = 2000,
        meta_model: str = None,
    ):
        self.n = n
        self.model = model
        self.temperature = temperature
        self.max_tokens = max_tokens
        self.meta_model = meta_model or model
        self.client = anthropic.AsyncAnthropic(api_key=os.environ["ANTHROPIC_API_KEY"])

    async def _sample_one(self, prompt: str, index: int) -> dict:
        response = await self.client.messages.create(
            model=self.model,
            max_tokens=self.max_tokens,
            temperature=self.temperature,
            messages=[{"role": "user", "content": prompt}],
        )
        return {"index": index, "content": response.content[0].text}

    async def _synthesize(self, responses: list[dict], custom_meta_prompt: str = None) -> str:
        meta_template = custom_meta_prompt
        if not meta_template and META_PROMPT_PATH.exists():
            meta_template = META_PROMPT_PATH.read_text()
        else:
            meta_template = "You are a structural architect synthesizing multiple AI responses."

        formatted = "\n\n---\n\n".join(
            f"### Response {r['index']}\n\n{r['content']}" for r in responses
        )

        meta_prompt = f"{meta_template}\n\n---\n\n## The {self.n} responses to synthesize:\n\n{formatted}"

        response = await self.client.messages.create(
            model=self.meta_model,
            max_tokens=self.max_tokens * 2,
            temperature=0.3,  # Low temperature for synthesis
            messages=[{"role": "user", "content": meta_prompt}],
        )
        return response.content[0].text

    def run(self, prompt: str, custom_meta_prompt: str = None) -> dict:
        """Synchronous wrapper for the async pipeline."""
        return asyncio.run(self._run_async(prompt, custom_meta_prompt))

    async def _run_async(self, prompt: str, custom_meta_prompt: str = None) -> dict:
        print(f"[PRISM] Starting N-Sampling (n={self.n})...")
        start = time.time()

        # Step 1: Parallel sampling
        tasks = [self._sample_one(prompt, i + 1) for i in range(self.n)]
        responses = await asyncio.gather(*tasks)
        print(f"[PRISM] {self.n} responses received ({(time.time()-start)*1000:.0f}ms)")

        # Step 2: Meta-agent synthesis
        print("[PRISM] Running Meta-Agent synthesis...")
        synthesis = await self._synthesize(responses, custom_meta_prompt)

        total_ms = (time.time() - start) * 1000
        print(f"[PRISM] Synthesis complete ({total_ms:.0f}ms total)")

        return {
            "synthesis": synthesis,
            "responses": responses,
            "metadata": {
                "n": self.n,
                "model": self.model,
                "temperature": self.temperature,
                "total_ms": total_ms,
            },
        }


if __name__ == "__main__":
    pipeline = PRISMPipeline(n=3)  # n=3 for demo speed

    prompt = """
    Explain the most underrated recovery technique in strength training,
    with physiological mechanisms and practical implementation.
    Be specific, avoid common knowledge.
    """

    result = pipeline.run(prompt)

    print("\n" + "="*60)
    print("PRISM SYNTHESIS (Meta-Agent Output)")
    print("="*60)
    print(result["synthesis"])
    print("\n" + "="*60)
    print(f"Total: {result['metadata']['total_ms']:.0f}ms | N={result['metadata']['n']}")
